Add support for file-based configuration

mtlewis opened the pull request:
## Description

This PR adds support for loading lockfile-lint configuration from a file, rather than supplying it on the command line. It uses [cosmiconfig](https://www.npmjs.com/package/cosmiconfig), which gives us a bunch of sensible configuration file support for free. Command-line options always override options specified in config files.

## Types of changes

<!--- What types of changes does your code introduce? Put an `x` in all the boxes that apply: -->

- [ ] Bug fix (non-breaking change which fixes an issue)
- [x] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to change)

## Related Issue

Welp - I didn't notice that only PRs with open issues would be accepted before I worked on this, so I created one (#74). Sorry for not quite following the guidelines!

## Motivation and Context

See #74.

## How Has This Been Tested?

So far, I've only tested this manually. I'll be happy to circle back and add tests if there's a positive reaction to #74.

## Checklist:

- [x] I have updated the documentation (if required).
- [x] I have read the **CONTRIBUTING** document.
- [x] I have added tests to cover my changes.
- [x] All new and existing tests passed.
- [x] I added a picture of a cute animal cause it's fun

![b88475796z1_20161206174812_000g7ge7mhd2-0-4el676zv61zhvnrpcn2_ct677x380](https://user-images.githubusercontent.com/542836/75998116-39a16400-5ef8-11ea-8e68-287623e23e38.jpg)

lirantal commented:
This is wonderful @mtlewis, thanks for working on this ❤️
Lovely picture too!

I'll review it shortly

lirantal commented on packages/lockfile-lint/README.md:
```diff
@@ -78,6 +78,19 @@ lockfile-lint --path yarn.lock --allowed-hosts yarn github.com --validate-https
 
+# File-Based Configuration
+
+Lockfile-lint uses [cosmiconfig]() for configuration file support. This means you can configure the above options via (in order of precedence):
```
probably want to add the npmjs link for`cosmiconfig`

mtlewis replied:
😁 oops, sorry! I _meant_ to...

mtlewis replied:
Fixed in 3b61e37

lirantal commented on packages/lockfile-lint/src/config.js:
```diff
@@ -0,0 +1,30 @@
+
+let fileConfig
+if (cosmiconfigResult) {
+  fileConfig = validateConfig(cosmiconfigResult.config)
```
@mtlewis is this a synchronous operation? what happens if it throws?

mtlewis replied:
Yes, it's synchronous since we're using `cosmiconfigSync`. I hadn't spotted that it could throw though, thanks for pointing it out. Newest version of the code now handles errors, ignoring file-based config and logging an error.

lirantal commented:
@mtlewis I'm happy to land this.
Are you able to add the relevant tests to match up with code coverage requirement? that's why the build is failing now.

mtlewis commented:
Hey @lirantal, thanks for your review! I've added some tests. I also did a bit of refactoring of the PR, since the way it was before made it harder to retain the argument validation that's configured in yargs. Instead, we now load the cosmiconfig first, then pass it into yargs as the base config. This allows us to keep the configuration rules in yargs. I've also added unit tests :+1:.

mtlewis commented:
By the way - I'm not sure coverage is currently working in cli.js. I believe I've covered all the new code in cli.js in tests, but both before and after this change, the detected coverage in cli.js is 0. I suspect this could be something to do with the fact that the tests for cli.js are currently executing in a separate process.

lirantal commented:
Yep, I see what you mean there about the cli.js not being covered in unit tests (and indeed in a different process they aren't being counted for).

What if you move this code https://github.com/lirantal/lockfile-lint/blob/0a90e4291cbf1d41a9246d5a38f7567f11781cc5/packages/lockfile-lint/src/cli.js#L5-L20 to its own `config.js` module, add unit test coverage to it, and then require it in cli.js?

Tip: to better understand what lines of code are missing coverage, you can run `yarn run coverage:view`

mtlewis commented:
Hey @lirantal thanks for the feedback and suggestion - how about this: 76df197? Since the whole of cli.js was previously uncovered, I figured it might be good to refactor the whole thing so that it's testable without launching a child process, and then separate out the testing of the option parsing/validation from the cli tests in cli.js.

lirantal commented:
Ok, taking a look ;-)

lirantal commented:
this looks overall good, thanks @mtlewis 🙏
